src/pal/pal_lower_kernel_arguments_pass.h

```diff
@@ -0,0 +1,26 @@
+// Copyright (c) 2023 Advanced Micro Devices, Inc. All rights reserved.
+
+#ifndef PAL_PLATFORM_LOWER_KERNEL_ARGUMENTS_H
+#define PAL_PLATFORM_LOWER_KERNEL_ARGUMENTS_H
+
+#include <llvm/IR/PassManager.h>
+
+/// This pass replaces accesses to kernel arguments with loads from offsets from a manually supplied buffer
+/// containing these arguments. The pointer to this buffer is expected to be prepopulated into specific sgprs
+/// by the PALPlatform.
+///
+/// This pass is an almost 1:1 replicate of the AMDGPULowerKernelArguments pass
+/// (llvm-project/llvm/lib/Target/AMDGPU/AMDGPULowerKernelArguments.cpp)
```

Comment:
Is there a reason why we wouldn't just use the LLVM pass?

Reply:
Yes. The LLVM pass relies on `gcn_dispatch_ptr()`, which relies on the convention that every HSA dispatch is triggered by an AQL packet that provides a buffer with the kernel arguments. We don't have all of that luxury with PAL out of the box, so we build up the buffer with the kernel arguments manually. The problem with this is, that the pointer to that buffer is placed in SGPRs, which is not what the LLVM pass expects. So, the change we had to make is the code that's generated to get the pointer to the kernargs buffer. Since this is "buried" inside the function of the original LLVM pass I didn't see a better option than moving the pass here and adding our modifications. (See comments in the cpp file)